cmd/zoekt-sourcegraph-indexserver/backoff_test.go

```diff
@@ -0,0 +1,249 @@
+package main
+
+import (
+	"github.com/sourcegraph/log/logtest"
+	"testing"
+	"time"
+)
+
+func TestQueue_BackoffOnFail(t *testing.T) {
+	backoffDuration := 1 * time.Millisecond
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	bumpTime := time.Now()
+	queue.Bump([]uint32{opts.RepoID})
+
+	// item is disallowed from being pushed to heap during backoff period
+	if item, ok := queue.Pop(); ok {
+		qi := queue.items[item.RepoID]
+		if qi.backoff.backoffUntil.Before(bumpTime) {
+			t.Errorf("backoffDuration already passed before first attempt to push item to heap in Bump(). Increase backoffDuration for the Queue. backoffDuration: %s. maxBackoffDuration: %s.",
+				backoffDuration, maxBackoffDuration)
+		} else {
+			t.Fatal("queue should be empty")
+		}
+	}
+}
+
+func TestQueue_BackoffAllowAfterDuration(t *testing.T) {
+	backoffDuration := 1 * time.Millisecond
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	time.Sleep(backoffDuration * 2)
+
+	queue.Bump([]uint32{opts.RepoID})
+
+	if _, ok := queue.Pop(); !ok {
+		t.Fatal("queue should no longer be empty after waiting for longer than the backoff duration and then bumping index options")
+	}
+}
+
+func TestQueue_ResetBackoffUntil(t *testing.T) {
+	backoffDuration := 1 * time.Hour
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
```

Comment:
should we not do a `Bump` or `AddOrUpdate` just before this to ensure we are running backoff logic?

Reply:
In this case I'm just checking that nothing is pushed to the queue after `SetIndexed()`, which I do in a lot of tests as a sanity check. This check is not critical to the backoff logic but just a way to ensure we catch a regression where `SetIndexed()` pushes to the queue (atm it'll either ensure order is correct or remove a repo; it will never add). I'm ok with removing it.

I skip checking the bump or addOrUpdate after the failure since I check that in the backoffOnFail test. This test will just test the behavior after the successful indexing and resetting the backoff, so I wait until I submit indexStateSuccess on line 86 and then bump and assert there